### DISCUSSION TITLE
BadParentNodeIdInvalid when importing FDI nodeset

### DISCUSSION THREAD

**Discussion by siGrau** [FreeOpcUa/python-opcua#1266] [Q&A] [answered]
Hi,
I´m trying to run an OPC UA Server with the FDI NodeSet.
Importing the Opc.Ua.Fdi5.NodeSet2.xml fails with the following error:
```bash
Parsing value of type 'QualifiedName' not implemented
failure adding node NodeData(nodeid:NumericNodeId(ns=1;i=92))
"The parent node id does not to refer to a valid node."(BadParentNodeIdInvalid)
Stapelüberwachung:
 >  File "C:\Users\simon\projects\server\FDITest\FDITest\FDITest.py", line 28, in <module>
 >    server.import_xml("ua-nodesets/FDI/Opc.Ua.Fdi5.NodeSet2.xml")
"__main__" geladen
"runpy" geladen
Traceback (most recent call last):
  File "C:\Program Files\Python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Program Files\Python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "c:\program files (x86)\microsoft visual studio\2019\community\common7\ide\extensions\microsoft\python\core\debugpy\__main__.py", line 45, in <module>
    cli.main()
  File "c:\program files (x86)\microsoft visual studio\2019\community\common7\ide\extensions\microsoft\python\core\debugpy/..\debugpy\server\cli.py", line 430, in main
    run()
  File "c:\program files (x86)\microsoft visual studio\2019\community\common7\ide\extensions\microsoft\python\core\debugpy/..\debugpy\server\cli.py", line 267, in run_file
    runpy.run_path(options.target, run_name=compat.force_str("__main__"))
  File "C:\Program Files\Python37\lib\runpy.py", line 263, in run_path
    Der Thread 'MainThread' (0x1) hat mit Code 0 (0x0) geendet.
pkg_name=pkg_name, script_name=fname)
  File "C:\Program Files\Python37\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Program Files\Python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\simon\projects\server\FDITest\FDITest\FDITest.py", line 28, in <module>
    server.import_xml("ua-nodesets/FDI/Opc.Ua.Fdi5.NodeSet2.xml")
  File "C:\Users\simon\AppData\Roaming\Python\Python37\site-packages\opcua\server\server.py", line 510, in import_xml
    return importer.import_xml(path, xmlstring)
  File "C:\Users\simon\AppData\Roaming\Python\Python37\site-packages\opcua\common\xmlimporter.py", line 67, in import_xml
    node = self._add_node_data(nodedata)
  File "C:\Users\simon\AppData\Roaming\Python\Python37\site-packages\opcua\common\xmlimporter.py", line 94, in _add_node_data
    node = self.add_method(nodedata)
  File "C:\Users\simon\AppData\Roaming\Python\Python37\site-packages\opcua\common\xmlimporter.py", line 360, in add_method
    res[0].StatusCode.check()
  File "C:\Users\simon\AppData\Roaming\Python\Python37\site-packages\opcua\ua\uatypes.py", line 218, in check
    raise UaStatusCodeError(self.value)
opcua.ua.uaerrors._auto.BadParentNodeIdInvalid: "The parent node id does not to refer to a valid node."(BadParentNodeIdInvalid)
```

**To Reproduce**<br />
The following code is just the server-import-robot-nodeset.py example with Opc.Ua.Fdi5.NodeSet2.xml instead of the robot nodeset:
```
import sys
sys.path.insert(0, "..")
import time


from opcua import ua, Server


if __name__ == "__main__":

    # setup our server
    server = Server()
    server.set_endpoint("opc.tcp://0.0.0.0:4840/freeopcua/server/")

    # setup our own namespace, not really necessary but should as spec
    uri = "http://examples.freeopcua.github.io"
    idx = server.register_namespace(uri)

    # get Objects node, this is where we should put our nodes
    objects = server.get_objects_node()

    # populating our address space
    myobj = objects.add_object(idx, "MyObject")
    myvar = myobj.add_variable(idx, "MyVariable", 6.7)
    myvar.set_writable()    # Set MyVariable to be writable by clients

    server.import_xml("ua-nodesets/DI/Opc.Ua.Di.NodeSet2.xml")
    server.import_xml("ua-nodesets/FDI/Opc.Ua.Fdi5.NodeSet2.xml")

    # starting!
    server.start()

    try:
        count = 0
        while True:
            time.sleep(1)
            count += 0.1
            myvar.set_value(count)
    finally:
        #close connection, remove subcsriptions, etc
        server.stop()
```

**Additional Info**<br />
Importing just the DI NodeSet works fine. In addition importing any other NodeSet which is built on the DI NodeSet like Robotics works fine,too.

**Version**<br />
Python-Version: 3.7.9
python-opcua Version: 0.98.12
DI NodeSet Version: 1.02.2 (the latest commit)
FDI5 NodeSet Version: 1.04.3 (the latest commit)

**Comment by vidyarao123** [comment]
I have faced a similar issue. The below lines are from the xml that I use to import. 


UAVariable NodeId="ns=1;i=4002" BrowseName="1:Tag1999"  ParentNodeID="i=85" DataType="Float" AccessLevel="3" UserAccessLevel="3">
    <DisplayName>Tag1999</DisplayName>
    <Description>tag1999</Description>
    <References> 
      <Reference ReferenceType="HasComponent" IsForward="false">i=85</Reference>
      <Reference ReferenceType="HasTypeDefinition">i=63</Reference> 
    </References>
    <Value> 
      <uax:Float>0.1</uax:Float>
    </Value>
  </UAVariable>

I noticed that I get "Bad parent ID ..." error mostly because of the typing mistake. I resolved by correcting at 2 places in my xml

1. Check the spelling of ParentNodeID
2. The "false" in line no. 6 (in the above xml) must be in lower case. If it is upper case, it doesnt accept.

I hope this helps 

Tha

**Comment by swamper123** [comment, marked as answer]
We had some issues in [opcua-asyncio](https://github.com/FreeOpcUa/opcua-asyncio), which is the main repo, since this is in maintainance mode, where such kind of issues should be cleared.
There was an issue with missing nodes in the UA set (since the part.xml files shouldn't be used anymore), problems with the addresspace (identical browsenames on same hierarchy level), up to missing- and unchecked (version dependencies) Nodes and references with the xml importer.
I recommened you to switch to the newer project. Depending on my timeschedule I may provide a PR for that (but not in near future).

**Reply by siGrau** [reply]
Thank you for your help, but I´ve had the same issue with opcua-asyncio and open62541. Do you think that there could be a mistake in the FDI nodeset itself?

**Reply by AndreasHeine** [reply]
could be! wouldnt be the first time ^^

**Reply by swamper123** [reply]
@siGrau Well your code seems to be okay from the order of importing etc.
Can you track down the Node which causes the error? 
Either there is an issue in the Nodeset (which it seems to be) or it can be caused by a self made error I caused in opcua-asyncio(fix incoming). 😅

**Reply by siGrau** [reply]
The following code is out of the Opc.Ua.Fdi5.NodeSet2.xml file and shows the node, which causes the BadParentNodeIdInvalid Error:

![FDI5_i=92](https://user-images.githubusercontent.com/52404207/108689073-ef463800-74f8-11eb-9cfa-7b3a2e4e185a.PNG)

The node is no other time mentioned which is weird, because the node is a method without a parent reference.

**Reply by swamper123** [reply]
That could be the problem. You may want to report that issue on mantis, the "Issue board" from OPC UA. https://mantis.opcfoundation.org/my_view_page.php

**Reply by siGrau** [reply]
Alright, thank you for your help!
